Removed trivial copy constructors on parameter classes to enable devi…

HouQiming opened the pull request:
Currently `cutlass::gemm::kernel::DefaultGemm<...>::GemmKernel` cannot be launched from device (CUDA Dynamic Parallelism) because its `Params` class indirectly contains two classes with user-defined constructors: `Coord` (base class of `GemmCoord`) and `TensorRef`.

This patch removes such copy constructors, enabling the kernels to be launched from device-side. These copy constructors are trivial anyway so removing them shouldn't affect functionality.

Test code that fails to build without patch and works with patch:

```C++
#include "cutlass/cutlass.h"
#include "cutlass/gemm/device/gemm.h"
#include "cutlass/gemm/kernel/default_gemm.h"
#include "cutlass/gemm/device/default_gemm_configuration.h"

using OperatorClass = cutlass::arch::OpClassSimt;
using ArchTag = cutlass::arch::Sm86;
using ThreadblockShape = cutlass::gemm::device::DefaultGemmConfiguration<
	OperatorClass, ArchTag, float, float, float, float>::ThreadblockShape;
/// Warp-level tile size (concept: GemmShape)
using WarpShape = cutlass::gemm::device::DefaultGemmConfiguration<
	OperatorClass, ArchTag, float, float, float, float>::WarpShape;
/// Instruction-level tile size (concept: GemmShape)
using InstructionShape = cutlass::gemm::device::DefaultGemmConfiguration<
	OperatorClass, ArchTag, float, float, float, float>::InstructionShape;
/// Epilogue output operator
using EpilogueOutputOp = cutlass::gemm::device::DefaultGemmConfiguration<
	OperatorClass, ArchTag, float, float, float, float>::EpilogueOutputOp;
using ThreadblockSwizzle = cutlass::gemm::threadblock::GemmIdentityThreadblockSwizzle<>;
static const int kStages = cutlass::gemm::device::DefaultGemmConfiguration<OperatorClass, ArchTag, float, float, float, float>::kStages;
/// Access granularity of A matrix in units of elements
static const int kAlignmentA = cutlass::gemm::device::DefaultGemmConfiguration<OperatorClass, ArchTag, float, float, float, float>::kAlignmentA;
/// Access granularity of B matrix in units of elements
static const int kAlignmentB = cutlass::gemm::device::DefaultGemmConfiguration<OperatorClass, ArchTag, float, float, float, float>::kAlignmentB;
static const bool kSplitKSerial = false;
template<typename LayoutA, typename LayoutB, typename LayoutC>
using CutlassGemmKernel = typename cutlass::gemm::kernel::DefaultGemm<
  float,
  LayoutA,
  kAlignmentA,
  float,
  LayoutB,
  kAlignmentB,
  float,
  LayoutC,
  float,
  OperatorClass,
  ArchTag,
  ThreadblockShape,
  WarpShape,
  InstructionShape,
  EpilogueOutputOp,
  ThreadblockSwizzle,
  kStages,
  kSplitKSerial,
  cutlass::gemm::device::DefaultGemmConfiguration<OperatorClass, ArchTag, float, float, float, float>::Operator
>::GemmKernel;

template<typename T>
void check(T result, char const *const func, const char *const file, int const line) {
	if (result) {
		fprintf(stderr, "CUDA error at %s:%d code=%d \"%s\" \n",
				file, line, static_cast<unsigned int>(result), func);
		// Make sure we call CUDA Device Reset before exiting
		abort();
	}
}
#define checkCudaErrors(val)           check((val), #val, __FILE__, __LINE__)

template<typename LayoutA, typename LayoutB, typename LayoutC>
__device__ void run(
	cutlass::gemm::GemmCoord problem_size,
	cutlass::TensorRef<float const, LayoutA> ref_A,
	cutlass::TensorRef<float const, LayoutB> ref_B,
	cutlass::TensorRef<float const, LayoutC> ref_C,
	cutlass::TensorRef<float, LayoutC> ref_D,
	typename EpilogueOutputOp::Params epilogue = EpilogueOutputOp::Params()
	//int split_k_slices=1
) {
	using TGemmKernel = CutlassGemmKernel<LayoutA, LayoutB, LayoutC>;
	ThreadblockSwizzle threadblock_swizzle;
	cutlass::gemm::GemmCoord grid_shape = threadblock_swizzle.get_tiled_shape(
		problem_size, 
		{ThreadblockShape::kM, ThreadblockShape::kN, ThreadblockShape::kK},
		1//split_k_slices
	);
	
	// Initialize the Params structure
	typename TGemmKernel::Params params{
		problem_size,
		grid_shape,
		ref_A.non_const_ref(),
		ref_B.non_const_ref(),
		ref_C.non_const_ref(),
		ref_D,
		epilogue,
		NULL //workspace
	};
	dim3 grid = threadblock_swizzle.get_grid_shape(params.grid_tiled_shape);
	dim3 block(TGemmKernel::kThreadCount, 1, 1);
	int smem_size = int(sizeof(typename TGemmKernel::SharedStorage));
	cutlass::Kernel<TGemmKernel><<<grid, block, smem_size >>> (params);
}

__global__ void gpu_master(float* a, float* b, float* c, int n) {
	int i = blockIdx.x * blockDim.x + threadIdx.x;
	a += i * n * n;
	b += i * n * n;
	c += i * n * n;
	run<cutlass::layout::RowMajor, cutlass::layout::RowMajor, cutlass::layout::RowMajor>(
		{n,n,n},
		{a, n},
		{b, n},
		{c, n},
		{c, n},
		{1.f, 0.f}
	);
}

int main() {
	float* a = NULL;
	float* b = NULL;
	float* c = NULL;
	int n = 128;
	int m = 7 * 32;
	cudaMallocHost((void**) & a, m * n * n * sizeof(float));
	cudaMallocHost((void**) & b, m * n * n * sizeof(float));
	cudaMallocHost((void**) & c, m * n * n * sizeof(float));
	cudaMemset(a, 0x3f, m * n * n * sizeof(float));
	cudaMemset(b, 0x3f, m * n * n * sizeof(float));
	cudaMemset(c, 0x3f, m * n * n * sizeof(float));
	gpu_master<<<7, 32 >>> (a, b, c, n);
	checkCudaErrors(cudaPeekAtLastError());
	float* cc = (float*)malloc(m * n * n * sizeof(float));
	cudaMemcpy(cc, c, m * n * n * sizeof(float), cudaMemcpyDeviceToHost);
	printf("%f %f %f\n", cc[0], cc[n * n + 1], cc[n * n * (m - 1) + n * n - 7]);
	return 0;
}
```

Tested with `-arch=compute_86 -code=sm_86 -rdc=true` on CUDA 11. Expected output:
```
71.4364 71.4364 71.4364
```

hwu36 commented:
Sorry for the late response.  I am just back from the vacation.

What about just changing `CUTLASS_HOST_DEVICE` to `__host__`?

HouQiming commented:
It didn't work, in addition to the original copy constructor error, nvcc
also reports:

/home/hqm/tp/cutlass/include/cutlass/gemm/kernel/gemm_splitk_parallel.h(98):
warning #20014-D: calling a __host__ function from a __host__ __device__
function is not allowed

On Tue, Dec 7, 2021 at 1:56 AM Haicheng Wu ***@***.***> wrote:

> Sorry for the late response. I am just back from the vacation.
>
> What about just changing CUTLASS_HOST_DEVICE to __host__?
>
> —
> You are receiving this because you authored the thread.
> Reply to this email directly, view it on GitHub
> <https://github.com/NVIDIA/cutlass/pull/366#issuecomment-987015648>, or
> unsubscribe
> <https://github.com/notifications/unsubscribe-auth/AB5MQF3KNW44XPTSYTUPARTUPT2NRANCNFSM5IXORE6A>
> .
> Triage notifications on the go with GitHub Mobile for iOS
> <https://apps.apple.com/app/apple-store/id1477376905?ct=notification-email&mt=8&pt=524675>
> or Android
> <https://play.google.com/store/apps/details?id=com.github.android&referrer=utm_campaign%3Dnotification-email%26utm_medium%3Demail%26utm_source%3Dgithub>.
>
>

HouQiming commented:
Hi, any update?

On Fri, Jan 7, 2022 at 10:23 AM github-actions[bot] <
***@***.***> wrote:

> This PR has been labeled inactive-30d due to no recent activity in the
> past 30 days. Please close this PR if it is no longer required. Otherwise,
> please respond with a comment indicating any updates. This PR will be
> labeled inactive-90d if there is no activity in the next 60 days.
>
> —
> Reply to this email directly, view it on GitHub
> <https://github.com/NVIDIA/cutlass/pull/366#issuecomment-1007088178>, or
> unsubscribe
> <https://github.com/notifications/unsubscribe-auth/AB5MQF552CK4UW5PZHGFIV3UUZFA5ANCNFSM5IXORE6A>
> .
> Triage notifications on the go with GitHub Mobile for iOS
> <https://apps.apple.com/app/apple-store/id1477376905?ct=notification-email&mt=8&pt=524675>
> or Android
> <https://play.google.com/store/apps/details?id=com.github.android&referrer=utm_campaign%3Dnotification-email%26utm_medium%3Demail%26utm_source%3Dgithub>.
>
> You are receiving this because you authored the thread.Message ID:
> ***@***.***>
>

HouQiming commented:
Hi?

On Sun, Feb 6, 2022 at 12:06 PM github-actions[bot] <
***@***.***> wrote:

> This PR has been labeled inactive-30d due to no recent activity in the
> past 30 days. Please close this PR if it is no longer required. Otherwise,
> please respond with a comment indicating any updates. This PR will be
> labeled inactive-90d if there is no activity in the next 60 days.
>
> —
> Reply to this email directly, view it on GitHub
> <https://github.com/NVIDIA/cutlass/pull/366#issuecomment-1030747481>, or
> unsubscribe
> <https://github.com/notifications/unsubscribe-auth/AB5MQF3ICN7Q6JZK3YQPHNLUZXXSXANCNFSM5IXORE6A>
> .
> Triage notifications on the go with GitHub Mobile for iOS
> <https://apps.apple.com/app/apple-store/id1477376905?ct=notification-email&mt=8&pt=524675>
> or Android
> <https://play.google.com/store/apps/details?id=com.github.android&referrer=utm_campaign%3Dnotification-email%26utm_medium%3Demail%26utm_source%3Dgithub>.
>
> You are receiving this because you authored the thread.Message ID:
> ***@***.***>
>

hwu36 commented:
This one is mostly okay.  Will merge after some testing.  Sorry for the delay.

HouQiming commented:
No problem, thanks!

On Sun, Feb 6, 2022 at 12:39 PM Haicheng Wu ***@***.***>
wrote:

> This one is mostly okay. Will merge after some testing. Sorry for the
> delay.
>
> —
> Reply to this email directly, view it on GitHub
> <https://github.com/NVIDIA/cutlass/pull/366#issuecomment-1030750678>, or
> unsubscribe
> <https://github.com/notifications/unsubscribe-auth/AB5MQF3W74WP6LT25RQERPLUZX3Q5ANCNFSM5IXORE6A>
> .
> Triage notifications on the go with GitHub Mobile for iOS
> <https://apps.apple.com/app/apple-store/id1477376905?ct=notification-email&mt=8&pt=524675>
> or Android
> <https://play.google.com/store/apps/details?id=com.github.android&referrer=utm_campaign%3Dnotification-email%26utm_medium%3Demail%26utm_source%3Dgithub>.
>
> You are receiving this because you authored the thread.Message ID:
> ***@***.***>
>

hwu36 commented:
Hi,

Sorry for the delay.  I tried your PR, but it seems your change fails example 06.  Seems these copy ctors are used.

HouQiming commented:
Thanks for pointing that out! After closer check, example 6 uses one constructor I removed as a converting constructor. But the same constructor is mistaken as a copy constructor in different instantiations.

I added a SFINAE mechanism so that the constructor is only created when it's converting, i.e., for example 6, and not for device code.

hwu36 commented on include/cutlass/tensor_ref.h:
```diff
@@ -219,9 +219,12 @@ class TensorRef {
-    NonConstTensorRef const &ref              ///< TensorRef to non-const data
+    NonConstTensorRef const &ref,             ///< TensorRef to non-const data
+    ///SFINAE trick to avoid creating a copy-constructor when Element_ is already non-const
+    _Magic magic = (typename std::enable_if< ! std::is_same<NonConstTensorRef, TensorRef<Element_, Layout_> >::value, _Magic>::type)0
```
std:: cannot be used with nvrtc.  Is it possible to use platform:: instead here

HouQiming commented:
Yes. Just updated.

On Mon, Feb 21, 2022 at 2:34 PM Haicheng Wu ***@***.***>
wrote:

> ***@***.**** commented on this pull request.
> ------------------------------
>
> In include/cutlass/tensor_ref.h
> <https://github.com/NVIDIA/cutlass/pull/366#discussion_r810813220>:
>
> >    CUTLASS_HOST_DEVICE
>    TensorRef(
> -    NonConstTensorRef const &ref              ///< TensorRef to non-const data
> +    NonConstTensorRef const &ref,             ///< TensorRef to non-const data
> +    ///SFINAE trick to avoid creating a copy-constructor when Element_ is already non-const
> +    _Magic magic = (typename std::enable_if< ! std::is_same<NonConstTensorRef, TensorRef<Element_, Layout_> >::value, _Magic>::type)0
>
> std:: cannot be used with nvrtc. Is it possible to use platform:: instead
> here?
>
> —
> Reply to this email directly, view it on GitHub
> <https://github.com/NVIDIA/cutlass/pull/366#pullrequestreview-888246697>,
> or unsubscribe
> <https://github.com/notifications/unsubscribe-auth/AB5MQF5GNJSBOFTYVTT5TLDU4HMG5ANCNFSM5IXORE6A>
> .
> Triage notifications on the go with GitHub Mobile for iOS
> <https://apps.apple.com/app/apple-store/id1477376905?ct=notification-email&mt=8&pt=524675>
> or Android
> <https://play.google.com/store/apps/details?id=com.github.android&referrer=utm_campaign%3Dnotification-email%26utm_medium%3Demail%26utm_source%3Dgithub>.
>
> You are receiving this because you authored the thread.Message ID:
> ***@***.***>
>

hwu36 commented:
It seems example 18 fails on Ampere with compute-sanitizer now.

HouQiming commented:
I reverted to commit 6b69c79ac3b208e22db1c65c00865be287533554 (my base
revision) and example 18 still crashed. So it doesn't seem to be my patch's
fault. I'm running it on a RTX3070.

What's that example supposed to do?


On Tue, Feb 22, 2022 at 9:53 PM Haicheng Wu ***@***.***>
wrote:

> It seems example 18 fails on Ampere with compute-sanitizer now.
>
> —
> Reply to this email directly, view it on GitHub
> <https://github.com/NVIDIA/cutlass/pull/366#issuecomment-1047819638>, or
> unsubscribe
> <https://github.com/notifications/unsubscribe-auth/AB5MQF3PZERZZYB7QL6YANTU4OIPNANCNFSM5IXORE6A>
> .
> Triage notifications on the go with GitHub Mobile for iOS
> <https://apps.apple.com/app/apple-store/id1477376905?ct=notification-email&mt=8&pt=524675>
> or Android
> <https://play.google.com/store/apps/details?id=com.github.android&referrer=utm_campaign%3Dnotification-email%26utm_medium%3Demail%26utm_source%3Dgithub>.
>
> You are receiving this because you authored the thread.Message ID:
> ***@***.***>
>

hwu36 commented:
> What's that example supposed to do?

That example just runs a affine-2 fp64 gemm on ampere.  affine-2 means no input and output data are consecutive.  The kernel cannot use vectorized load and store.

I can run the example on A100 without your patch.

hwu36 commented:
That example has some problem.  Let me take another look.

hwu36 commented on include/cutlass/coord.h:
```diff
@@ -94,14 +94,6 @@ struct Coord {
     }
   }
 
-  /// Copy constructor
```
Do we need SFINAE here too instead of deleting it?

hwu36 commented:
I fixed affine2 issues.

HouQiming commented:
I don't think so. The Coord constructor is always a copy constructor. The
TensorRef constructor needs SFINAE because it's sometimes a copy
constructor and sometimes a converting one.


On Sat, Feb 26, 2022 at 9:10 AM Haicheng Wu ***@***.***>
wrote:

> ***@***.**** commented on this pull request.
> ------------------------------
>
> In include/cutlass/coord.h
> <https://github.com/NVIDIA/cutlass/pull/366#discussion_r815244118>:
>
> > @@ -94,14 +94,6 @@ struct Coord {
>      }
>    }
>
> -  /// Copy constructor
>
> Do we need SFINAE here too instead of deleting it?
>
> —
> Reply to this email directly, view it on GitHub
> <https://github.com/NVIDIA/cutlass/pull/366#pullrequestreview-894347874>,
> or unsubscribe
> <https://github.com/notifications/unsubscribe-auth/AB5MQFYSCK4ECVW7TY4UEOLU5AR7NANCNFSM5IXORE6A>
> .
> Triage notifications on the go with GitHub Mobile for iOS
> <https://apps.apple.com/app/apple-store/id1477376905?ct=notification-email&mt=8&pt=524675>
> or Android
> <https://play.google.com/store/apps/details?id=com.github.android&referrer=utm_campaign%3Dnotification-email%26utm_medium%3Demail%26utm_source%3Dgithub>.
>
> You are receiving this because you authored the thread.Message ID:
> ***@***.***>
>

HouQiming commented:
Great!

On Sat, Feb 26, 2022 at 9:10 AM Haicheng Wu ***@***.***>
wrote:

> I fixed affine2 issues.
>
> —
> Reply to this email directly, view it on GitHub
> <https://github.com/NVIDIA/cutlass/pull/366#issuecomment-1051411914>, or
> unsubscribe
> <https://github.com/notifications/unsubscribe-auth/AB5MQFZDWROJ4MAAKSEE2F3U5ASB5ANCNFSM5IXORE6A>
> .
> Triage notifications on the go with GitHub Mobile for iOS
> <https://apps.apple.com/app/apple-store/id1477376905?ct=notification-email&mt=8&pt=524675>
> or Android
> <https://play.google.com/store/apps/details?id=com.github.android&referrer=utm_campaign%3Dnotification-email%26utm_medium%3Demail%26utm_source%3Dgithub>.
>
> You are receiving this because you authored the thread.Message ID:
> ***@***.***>
>
